tools/djxl_ng_main.cc

```diff
@@ -0,0 +1,299 @@
+              "JPEG output quality. Setting an output quality "
+              "implies --pixels_to_jpeg.");
+#endif
+
+#if JPEGXL_ENABLE_SJPEG
+// TODO(firsching): wire this up.
+DEFINE_bool(use_sjpeg, false, "use sjpeg instead of libjpeg for JPEG output");
+#endif
+
+// TODO(firsching): wire this up.
+DEFINE_bool(print_read_bytes, false, "print total number of decoded bytes");
+
+// TODO(firsching): wire this up.
+DEFINE_bool(quiet, false, "silence output (except for errors)");
+
+bool WriteFile(const char* filename, const std::vector<uint8_t> bytes) {
+  FILE* file = fopen(filename, "wb");
+  if (!file) {
+    fprintf(stderr,
+            "Could not open %s for writing\n"
+            "Error: %s",
+            filename, strerror(errno));
+    return false;
+  }
+  if (fwrite(bytes.data(), 1, bytes.size(), file) != bytes.size()) {
+    fprintf(stderr,
+            "Could not write to file\n"
+            "Error: %s",
+            strerror(errno));
+    return false;
+  }
+  if (fclose(file) != 0) {
+    fprintf(stderr,
+            "Could not close file\n"
+            "Error: %s",
+            strerror(errno));
+    return false;
+  }
+  return true;
+}
+
+int DecompressJxlReconstructJPEG(const char* filename,
+                                 std::vector<uint8_t>& jpeg_bytes,
+                                 JxlDecoderPtr dec,
+                                 JxlThreadParallelRunnerPtr runner,
+                                 JxlBasicInfo* info) {
+  FILE* file_in = fopen(filename, "rb");
+  if (!file_in) {
+    fprintf(stderr,
+            "Could not open %s for reading\n"
+            "Error: %s",
+            filename, strerror(errno));
+    return EXIT_FAILURE;
+  }
+  if (JXL_DEC_SUCCESS != JxlDecoderSetParallelRunner(dec.get(),
+                                                     JxlThreadParallelRunner,
+                                                     runner.get())) {
+    fprintf(stderr, "JxlEncoderSetParallelRunner failed\n");
+    return EXIT_FAILURE;
+  }
+
+  if (JXL_DEC_SUCCESS !=
+      JxlDecoderSubscribeEvents(
+          dec.get(), JXL_DEC_JPEG_RECONSTRUCTION | JXL_DEC_FULL_IMAGE)) {
+    fprintf(stderr, "JxlDecoderSubscribeEvents failed\n");
+    return EXIT_FAILURE;
+  }
+
+  // In how large chunks to read from the file.
+  const constexpr size_t kInputChunkSize = 65536;
+  const constexpr size_t kOutputChunkSize = 16384;
+  // Reading compressed JPEG XL input and decoding to pixels
+  std::vector<uint8_t> compressed;
+  size_t data_size = 0;
+  bool dec_successful = false;
+  bool can_reconstruct_jpeg = false;
+  std::vector<uint8_t> jpeg_data_chunk(kOutputChunkSize);
+  jpeg_bytes.resize(0);
+  for (;;) {
+    JxlDecoderStatus status = JxlDecoderProcessInput(dec.get());
+    if (status == JXL_DEC_ERROR) {
+      fprintf(stderr, "Failed to decode image\n");
+      break;
+    } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+      size_t remaining = JxlDecoderReleaseInput(dec.get());
```

Comment:
All of the logic here seems rather unnecessarily complicated, why don't we just read the whole file into memory?

Reply:
restructured to read the whole file into memory
It does make the logic easier.
I used the read file function from the `decode_oneshot.cc`. 
Right now djxl_ng_main.cc has some overlap of reading and writing files with cjxl_ng_main.cc. Not sure if it would make sense to have file_io in some central place for better reuse-ability.

Reply:
Makes sense, but we can do it in a followup